doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+
+Not all runtime errors can be catched by the contracts based design, for example, training binary classifiers with dataset that has three classes. Therefore, besides `contracts.py`, we need another mechanism to catch such errors and pass a user-friendly diagnostics messages to the user. The mechanism would be implemented in `diagnosics.py`.
+
+###### Golang
+
+The golang package `executor` should parse the diagnostics messages generated by `diagnostics.py` and return them to the users.
+
+In the remainder of the section, we say **diagnostics are required** where `diagnostics.py` should be involved to generate an error message.
+
+##### Intermediate Representations
+
+As described above, the proposed architecture requires the golang part to pass the *ir* of a SQLFlow program to the python part to execute the program. We propose to define `ir` as a protobuf file in `pkg/proto/ir.proto`. It should have at least the following messages and fields, each other detailed design doc could add necessary fields to the `Statement` message.
+
+```protobuf
+message Statement {
+  // `select` is the query for fetching data. For example,
+  // "select * from iris.train;"
+  optional string select = 2;
+  // `validation_select` is the query for fetching the validation data.
+  // For example, "select * from iris.val;".
+  optional string validation_select = 3;
+  // `model_image` is the name of the model's Docker image, for example, in the
+  // statement "TO TRAIN a_data_scientist/regressors:v0.2/MyDNNRegressor", the
+  // name "a_data_scientist/regressors:v0.2" is a Docker image.
+  optional string model_image = 4;
+  // `estimator` specifies the estimator type. For example, after parsing
+  // "select ... train DNNClassifier WITH ...", `estimator` will be
+  // "DNNClassifier".
+  optional string estimator = 5;
+  // `attributes` is a map of parsed attribute in the WITH Clause. For example,
+  // after parsing "select ... to train ... with train.epoch=1000,
+  // model.hidden_units = [10, 10]", the `attributes` will be
+  // {"train.epoch": "1000", "model.hidden_units": ""[10, 10]""}
+  optional map<string, string> attributes = 6;
+  repeated string label = 7;
+  message Columns {
+  // The COLUMN clause will be split into `column`. For example, in the COLUMN
+  // clause "COLUMN NUMERIC(sepal_length), INDICATOR(ID)" will be saved in
+  // `columns` as ["NUMERIC(sepal_length)", "INDICATOR(ID)"]
+    repeated string columns = 1;
+  }
+  // `columns` contain a map of string to `Columns` For multiple COLUMN clauses
+  // like "COLUMN ... FOR deep_feature, COLUMN ... FOR wide_feature". They will
+  // be parsed as {"deep_feature": [...], "wide_feature": [...]}. For single
+  // column clause like "column ...", "feature_columns" will be used as the
+  // default map key.
+  optional map<string, Columns> columns = 8;
+  // `model_save` specifies the saved model path in the INTO/USING clause.
+  optional bool model_save = 9;
+  enum Type {
+    RUN = 0;
```

Comment:
`RUN` here means `TO RUN` or normal SQL statement? How to distinguish them?

Reply:
> `RUN` here means `TO RUN` or normal SQL statement? How to distinguish them?

It'' be fixed in  the next commit.

Reply:
It's done.